### PR TITLE
Start adding docs to readme

### LINKED ISSUE #25: Remove `value` of Value type
Matchpy sets default `name` attribute we should use

### PR DESCRIPTION
**Pull request by saulshanabrook**
Thanks to a helpful discussion with @dcharbon on Friday, I have started outlining how this system works. 

- [x] Finish writing
- [ ] Think about if we should change names to make them more math (category theory)-ey. 
- [x] Add `replace_scan`
- [x] Test readme code
- [ ] Make notebook instead

Fixes #25

### CODE REVIEW

**Comment by saulshanabrook** [comment]
In this process, I realize I can add typing and register an operation very cleanly as a function!

Going from this:

```
class GetItem(matchpy.Operation):
    name = "GetItem"
    arity = matchpy.Arity(1, True)

# then in .pyi file
def GetItem(seq: CArray) -> CGetitem:
    ...
```

To this:

```python
@operation
def GetItem(length: CContent, getitem: CGetitem) -> CArray:
    ...
```

And `operation is pretty simple as well:

```
T = typing.TypeVar("T", bound=typing.Callable)


def operation(fn: T) -> T:
    """
    Register a matchpy operation for a function. 

    The function should have a fixed number of args and one return.

    This can also be done manually, by typing the result of `matchpy.Operation.new`
    as a callable, but this is more fluent.
    """
    sig = inspect.signature(fn)
    return matchpy.Operation.new(fn.__name__, matchpy.Arity(len(sig.parameters), True))
```

**Comment by saulshanabrook** [comment]
Trick is understanding that all this is just for compile time checking, not runtime.

**Comment by saulshanabrook** [comment]
OK this should be ready to merge. I have added some tests, but no CI yet. The tests just run some notebooks, run mypy and verify the readme works.

**Comment by saulshanabrook** [comment]
Closes #25

**Comment by saulshanabrook** [comment]
CI is working now on this and it is ready to merge.

File: uarray/uarray/core.py [hunk 1]
```diff
@@ -1,281 +1,280 @@
...
-register(Length(Sequence(w.length, w._)), lambda _, length: length)
+class CCallableBinary(typing.Generic[RET, ARG1, ARG2]):
+    pass
 
 
-class Scalar(matchpy.Operation):
-    """
-    Scalar(content)
-    """
+CGetitem = CCallableUnary[CArray, CContent]
```
Comment (hameerabbasi) [inline uarray/uarray/core.py]: Is it possible to rename this to `CGetItem`?

Comment (saulshanabrook) [reply uarray/uarray/core.py]: Yes :) Thanks for catching this.

**Comment by saulshanabrook** [comment]
@hameerabbasi good idea, I moved the types into separate files. How is that?

**Comment by hameerabbasi** [comment]
It's better, I was also wondering about the naming convention, what does the `C` signify? Is it necessary to have it? Python just has `PascalCase` convention, usually. Are you following the C++ way of prefacing type names with `C`?

It'd be nice to have all of them in a simple `typing` module, and we can make the rest of them private. Another way is to have them close to where they should be.

Examples of both: `typing.List` and `collections.Iterable`.

**Comment by saulshanabrook** [comment]
C stands for Category. All the names will need to changed, i have gotten feedback from a couple people that the resemblance to Pythons naming (GetItem, Sequence) is confusing. 

At least at the moment within this PR it is internally consistent that all compile time types are prefixed with C.

I expect the final look of how we write these to change. 

I'll add a few lines to the readme to explain how they are currently.

**Comment by hameerabbasi** [comment]
Perhaps have a `Category` base class? It could come in useful either way.

**Comment by saulshanabrook** [comment]
Sounds good. Added that.

**Comment by saulshanabrook** [comment]
I am gonna merge this in so that I can share the current state of the repo with others for feedback more easily.

**Pull request merged**